### PR TITLE
HH:MM:SS time formatting bug in crc-interactive

### PR DESCRIPTION
**Pull request by yassinkhalifa**
It seems when entering the decimal value for the parameter "-t" or "--time" in crc-interactive it's formatted in the srun_dict as "{}:00" which makes it interpreted as "MM:SS" but if using the string format "01:00" and it's formatted in the srun_dict as "{}:00" then it's interpreted as "HH:MM:SS" which is the necessary format for "srun". I added a possible fix, please take a look and feel free to discard it and fix it in any other way because I am sure this is not the best way to fix it, I was just trying to expose where the error comes from.

### CODE REVIEW

File: apps/crc_interactive.py [hunk 1]
```diff
@@ -87,10 +87,12 @@ def _validate_arguments(self, args: Namespace) -> None:
...
         if args.time.isdecimal():
             check_time = int(args.time)
+            args.time = '{}:00:00'.format(int(args.time))
         else:
             check_time = (
-                datetime.strptime(args.time, '%H:%M').hour +
-                datetime.strptime(args.time, '%H:%M').minute / 60)
+                datetime.strptime(args.time, '%H:%M:%S').hour +
+                datetime.strptime(args.time, '%H:%M:%S').minute / 60 +
+                datetime.strptime(args.time, '%H:%M:%S').minute / 3600)
```
Comment (djperrefort) [inline apps/crc_interactive.py:95]: I think this should be `datetime.strptime(args.time, '%H:%M:%S').seconds`:

```python
            datetime_obj = datetime.strptime(args.time, '%H:%M:%S')
            check_time = datetime_obj.hour + datetime_obj.minute / 60 + datetime_obj.second / 3600
```

**Comment by yassinkhalifa** [comment]
I totally agree that users don't really need the seconds part in time. If you want, I can just change that back to how it was before and just leave the decimal formatting line that I added (just modify it to be HH:MM which will be amended with seconds in srun_dict).

**Comment by djperrefort** [comment]
Sounds good to me.

**Comment by yassinkhalifa** [comment]
That's done now!

**Review by djperrefort — COMMENTED** [review]
I'd prefer to avoid mutating arguments to a function call, particularly a validation function. It violates the principle of least astonishment and can cause unexpected behavior down the line.

Proposition:

Why don't we move the time parsing logic into the commandline parser itself. We can define a top level function like:

```python
def parse_time(time_str: str) -> time:
    try:
        return time(*time_str.split(':'))

    except Exception:
        raise ArgumentTypeError(f'Could not parse time value {time_str}')
```

and have it cast automatically as follows:

```python
resource_args.add_argument('-t', '--time', type=parse_time, ...)
```

This makes life easier later in the code. Validation becomes much shorter:

```python
        check_time = args.time.hour + args.time.minute / 60
        if not self.min_time <= check_time <= self.max_time:
            ...
```

The only missing piece is updating the `create_srun_command` method to handle the new argument type. That should be fairly straight forward.

Thoughts?

**Comment by yassinkhalifa** [comment]
Yeah, this is much better! I can work on that.

**Comment by yassinkhalifa** [comment]
I added the suggested function parse_time and made it only return component #0 because we are always interested in  the hours component (the first element) which simplified the check_time part as you indicated and changed the formatting of default_time to just a decimal '1' so that it aligns with the output of parse_time. Please take a look and let me know if I misinterpreted any of your instructions.

File: apps/crc_interactive.py [hunk 2]
```diff
@@ -77,6 +77,13 @@ def __init__(self) -> None:
         additional_args.add_argument('-f', '--feature', help='specify a feature, e.g. `ti` for GPUs')
         additional_args.add_argument('-o', '--openmp', action='store_true', help='run using OpenMP style submission')
 
+    def parse_time(time_str: str) -> time:
```
Comment (djperrefort) [inline apps/crc_interactive.py:80]: My original suggestion was to define this as a top level function, but your implementation as method is equally valid. The only difference is that this method should be made static if included as a method:

```python
@staticmethod
def parse_time(time_str: str) -> time:
    ...
```

If left as is, the parent class will pass `self` as the first argument, which is not included in the signature (nor should it be in this case).

Leave this class as a method, but add the static tag.

Comment (djperrefort) [inline apps/crc_interactive.py:80]: Neither `time` nor `ArgumentTypeError` has been imported. Add an import statement for each of them:

```python
from argparse import Namespace, ArgumentTypeError
from datetime import time
```

Also the `datetime.datetime` class is no longer needed in the imports.

Comment (djperrefort) [inline apps/crc_interactive.py:80]: Add function docstring with `Args`, `Return`, and `Raises` sections. Here is an automatically generated example:

```python
    """Parse a string representation of time in 'HH:MM:SS' format and return a time object

    Args:
        time_str: A string representing time in 'HH:MM:SS' format.

    Returns:
        time: A time object representing the parsed time.

    Raises:
        ArgumentTypeError: If the input string is not in the correct format or cannot be parsed.
    """
```

File: apps/crc_interactive.py [hunk 3]
```diff
@@ -25,7 +25,7 @@ class CrcInteractive(BaseParser):
     min_time = 1  # Minimum limit on requested time in hours
     max_time = 12  # Maximum limit on requested time in hours
 
-    default_time = '1:00'  # Default runtime
+    default_time = '1'  # Default runtime
```
Comment (djperrefort) [inline apps/crc_interactive.py:28]: This needs to be the same object type as used by the parser argument definition. 

```python
default_time = time(1)
```

File: apps/crc_interactive.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ def __init__(self) -> None:
         resource_args = self.add_argument_group('Arguments for Increased Resources')
         resource_args.add_argument('-b', '--mem', type=int, default=self.default_mem, help='memory in GB')
         resource_args.add_argument(
-            '-t', '--time', default=self.default_time,
+            '-t', '--time', type=self.parse_time, default=self.default_time,
             help=f'run time in hours or hours:minutes [default: {self.default_time} hour] ')
```
Comment (djperrefort) [inline apps/crc_interactive.py:58]: Minor point, but lets drop the word hour from `[default: {self.default_time} hour]`. I think it reads more nicely with the fstring auto formatting of time objects.

You're welcome to ignore this suggestion if you disagree.

File: apps/crc_interactive.py [hunk 5]
```diff
@@ -85,12 +92,7 @@ def _validate_arguments(self, args: Namespace) -> None:
         """
 
         # Check wall time is between limits, enable both %H:%M format and integer hours
-        if args.time.isdecimal():
-            check_time = int(args.time)
-        else:
-            check_time = (
-                datetime.strptime(args.time, '%H:%M').hour +
-                datetime.strptime(args.time, '%H:%M').minute / 60)
+        check_time = int(args.time)
```
Comment (djperrefort) [inline apps/crc_interactive.py:95]: This type cast does not work. Python cannot automatically cast from time objects to integers. You need to handle hours, minutes, and seconds manually.

```
check_time = args.time.hour + args.time.minute / 60 + args.time.second / 3600
```

File: apps/crc_interactive.py [hunk 6]
```diff
@@ -122,7 +124,7 @@ def create_srun_command(self, args: Namespace) -> str:
         srun_dict = {
             'partition': '--partition={}',
             'num_nodes': '--nodes={}',
-            'time': '--time={}:00',
+            'time': '--time={}:00:00',
```
Comment (djperrefort) [inline apps/crc_interactive.py:127]: This creates `--time=01:00:00:00:00`. The correct form is `'time': '--time={}'`. Time objects automatically format to strings as `HH:MM:SS`.

**Review by djperrefort — CHANGES_REQUESTED** [review]
I've highlighted a mix of style and functional suggestions.

**Comment by yassinkhalifa** [comment]
I went through all the requested changes and completed them. Please let me know if there is anything else I need to change.

**Review by djperrefort — APPROVED** [review]

**Comment by djperrefort** [comment]
@yassinkhalifa please merge this in when you are ready

**Pull request merged**